const generic: unclear why "type annotations needed" error

I'm experimenting with replacing typenum with const generics, and replacing constraints involving constant generics being compared to other constant generics. I have gotten a little further recently, probably because of the recent merging of lazy normalization PRs. I have just stumbled upon a new issue.
 
```rust
#![feature(const_generics)]

pub struct IsLessOrEqual<const LHS: u32, const RHS: u32>;
pub struct Condition<const CONDITION: bool>;
pub trait True {}

impl<const LHS: u32, const RHS: u32> True for IsLessOrEqual<LHS, RHS> where
    Condition<{ LHS <= RHS }>: True
{
}
impl True for Condition<true> {}

struct S<const I: u32, const J: u32>;
impl<const I: u32, const J: u32> S<I, J>
where
    IsLessOrEqual<I, 8>: True,
    IsLessOrEqual<J, 8>: True,
    IsLessOrEqual<{ 8 - I }, { 8 - J }>: True,
    // Condition<{ 8 - I <= 8 - J }>: True,
{
    fn print() {
        println!("I {} J {}", I, J);
    }
}
```
This gives the error
```
error[E0283]: type annotations needed
  --> src/lib.rs:16:26
   |
5  | pub trait True {}
   | -------------- required by this bound in `True`
...
16 |     IsLessOrEqual<I, 8>: True,
   |                          ^^^^ cannot infer type for struct `IsLessOrEqual<I, 8u32>`
   |
   = note: cannot satisfy `IsLessOrEqual<I, 8u32>: True
```
In this specific case, replacing `IsLessOrEqual<{ 8 - I }, { 8 - J }>: True` with the commented-out `Condition<{ 8 - I <= 8 - J }>: True` makes the compilation work. However that is still not very robust: changing all three `IsLessOrEqual` constraints to the equivalent `Condition` constraints would again lead to compilation failure.

--- a/src/test/ui/const-generics/issues/issue-72787.rs
+++ b/src/test/ui/const-generics/issues/issue-72787.rs
@@ -1,0 +1,32 @@
+#![feature(const_generics)]
+#![allow(incomplete_features)]
+
+pub struct IsLessOrEqual<const LHS: u32, const RHS: u32>;
+pub struct Condition<const CONDITION: bool>;
+pub trait True {}
+
+impl<const LHS: u32, const RHS: u32> True for IsLessOrEqual<LHS, RHS> where
+    Condition<{ LHS <= RHS }>: True
+//~^ Error constant expression depends on a generic parameter
+{
+}
+impl True for Condition<true> {}
+
+struct S<const I: u32, const J: u32>;
+impl<const I: u32, const J: u32> S<I, J>
+where
+    IsLessOrEqual<I, 8>: True,
+    IsLessOrEqual<J, 8>: True,
+    IsLessOrEqual<{ 8 - I }, { 8 - J }>: True,
+//~^ Error constant expression depends on a generic parameter
+//~| Error constant expression depends on a generic parameter
+//~| Error constant expression depends on a generic parameter
+//~| Error constant expression depends on a generic parameter
+    // Condition<{ 8 - I <= 8 - J }>: True,
+{
+    fn print() {
+        println!("I {} J {}", I, J);
+    }
+}
+
+fn main() {}

--- a/src/test/ui/const-generics/issues/issue-72787.stderr
+++ b/src/test/ui/const-generics/issues/issue-72787.stderr
@@ -1,0 +1,42 @@
+error: constant expression depends on a generic parameter
+  --> $DIR/issue-72787.rs:9:32
+   |
+LL |     Condition<{ LHS <= RHS }>: True
+   |                                ^^^^
+   |
+   = note: this may fail depending on what value the parameter takes
+
+error: constant expression depends on a generic parameter
+  --> $DIR/issue-72787.rs:20:42
+   |
+LL |     IsLessOrEqual<{ 8 - I }, { 8 - J }>: True,
+   |                                          ^^^^
+   |
+   = note: this may fail depending on what value the parameter takes
+
+error: constant expression depends on a generic parameter
+  --> $DIR/issue-72787.rs:20:42
+   |
+LL |     IsLessOrEqual<{ 8 - I }, { 8 - J }>: True,
+   |                                          ^^^^
+   |
+   = note: this may fail depending on what value the parameter takes
+
+error: constant expression depends on a generic parameter
+  --> $DIR/issue-72787.rs:20:42
+   |
+LL |     IsLessOrEqual<{ 8 - I }, { 8 - J }>: True,
+   |                                          ^^^^
+   |
+   = note: this may fail depending on what value the parameter takes
+
+error: constant expression depends on a generic parameter
+  --> $DIR/issue-72787.rs:20:42
+   |
+LL |     IsLessOrEqual<{ 8 - I }, { 8 - J }>: True,
+   |                                          ^^^^
+   |
+   = note: this may fail depending on what value the parameter takes
+
+error: aborting due to 5 previous errors
+

